tfjs-layers/src/layers/nlp/samplers.ts

```diff
@@ -0,0 +1,226 @@
+      return promptDone.all().logicalNot().arraySync() as number === 1;
+    }
+
+    let iter = 0;
+    let logits: Tensor;
+    const maxIterations = maxLength - index;
+    while (iter <= maxIterations && cond(prompt, index)) {
+      // Compute the softmax distribution for the next token.
+      [logits, hiddenStates, cache] = next(prompt, cache, index);
+      const probabilities = softmax(logits.div(this.temperature));
+      // Compute next token.
+      let nextToken = this.getNextToken(probabilities);
+      // Don't overwrite anywhere mask is True.
+      nextToken = nextToken.cast(prompt.dtype);
+      nextToken = where(
+        mask.gather(index, 1),
+        prompt.gather(index, 1),
+        nextToken,
+      );
+      // Update the prompt with the next token.
+      nextToken = nextToken.expandDims(-1);
+      prompt = sliceUpdate(prompt, [0, index], nextToken);
+
+      index += 1;
+      iter += 1;
+    }
+    return prompt;
+  }
+
+  /**
+   * Get the next token based on given probability distribution over tokens.
+   * Subclasses must implement this method.
+   *
+   * @param probabilities the probability distribution for next
+   *    token over all vocab tokens.
+   */
+  abstract getNextToken(probabilities: Tensor): Tensor;
+
+  static fromConfig<T extends serialization.Serializable>(
+    cls: serialization.SerializableConstructor<T>,
+    config: serialization.ConfigDict): T {
+
+    return new cls(config);
+  }
+
+  getConfig(): serialization.ConfigDict {
+    return {temperature: this.temperature};
+  }
+}
+
+export interface TopKSamplerArgs extends SamplerArgs {
+  /**
+   * Integer, the `k` value of top-k.
+   * Defaults to 5.
+   */
+  k?: number;
+
+  /**
+   * Integer, the random seed.
+   * Defaults to `null`,
+   */
+  seed?: number;
+}
+
+/**
+ * Top-K Sampler class.
+ *
+ * This sampler implements top-k search algorithm. Briefly, top-k algorithm
+ * randomly selects a token from the tokens of top K probability, with
+ * selection chance determined by the probability.
+ *
+ * Examples:
+ * // TODO(pforderique): Add examples.
+ */
+export class TopKSampler extends Sampler {
+  k: number;
+  seed: number;
+
+  constructor(args: TopKSamplerArgs) {
+    super(args);
+    this.k = args.k;
+    this.seed = args.seed;
+  }
+
+  override getNextToken(probabilities: Tensor): Tensor {
+    // Filter out top-k tokens.
+    const k = 5;
+    const {values, indices} = topk(probabilities, k, false);
+
+    function randomSample(probabilities: Tensor): Tensor {
+      const probabilitiesArr = probabilities.arraySync() as number[][];
+      const samplesArr = [];
+
+      for (const probDistribution of probabilitiesArr) {
+        let sample =
+          Math.random() * probDistribution.reduce((a, b) => a + b, 0);
+        const sampleIdx = probDistribution.findIndex(val => {
+          return (sample -= val) <= 0;
+        });
+        samplesArr.push(sampleIdx);
+      }
```

Comment:
I feel like this can be done in `O(log(n))`, but that likely doesn't matter since the rest of the model takes much longer.
```suggestion
      // TODO(mattSoulanille): Investigate whether this should use binsearch.
      const probabilitiesArr = probabilities.arraySync() as number[][];
      const samplesArr = [];

      for (const probDistribution of probabilitiesArr) {
        let sample =
          Math.random() * probDistribution.reduce((a, b) => a + b, 0);
        const sampleIdx = probDistribution.findIndex(val => {
          return (sample -= val) <= 0;
        });
        samplesArr.push(sampleIdx);
      }
```

Reply:
sounds good! I'll leave the TODO - only thing is I'm not sure we can expect the distribution to be sorted in any way